src/io/ReadIO.hpp

```diff
@@ -0,0 +1,370 @@
+   *
+   * @param genericData The DataBlockGeneric structure containing the data and
+   * shape.
+   *
+   * @return A DataBlock structure containing the data and shape.
+   */
+  inline static DataBlock<DTYPE> fromGeneric(
+      const DataBlockGeneric& genericData)
+  {
+    auto result = DataBlock<DTYPE>(
+        std::any_cast<std::vector<DTYPE>>(genericData.data), genericData.shape);
+    return result;
+  }
+};
+
+/// Helper struct to check if a StorageObjectType is allowed. Used in static
+/// assert.
+template<StorageObjectType T>
+struct is_allowed_storage_object_type : std::false_type
+{
+};
+
+/// Helper struct to check if a StorageObjectType is allowed. Used in static
+/// assert.
+template<>
+struct is_allowed_storage_object_type<StorageObjectType::Dataset>
+    : std::true_type
+{
+};
+
+/// Helper struct to check if a StorageObjectType is allowed. Used in static
+/// assert.
+template<>
+struct is_allowed_storage_object_type<StorageObjectType::Attribute>
+    : std::true_type
+{
+};
+
+/**
+ * @brief Class for wrapping data objects (datasets or attributes) for reading
+ * data from a file
+ *
+ * @tparam OTYPE The type of object being wrapped defined via \ref
+ * AQNWB::Types::StorageObjectType
+ * @tparam VTYPE The data type of the values stored in the data object
+ */
+template<StorageObjectType OTYPE, typename VTYPE = std::any>
+class ReadDataWrapper
+{
+  // Embedded traits for compile time checking of allowed OTYPE for the class
+  // and methods
+private:
+  /**
+   * Embedded Trait to Check the OTYPE Enum Value at compile time to
+   * SFINAE (Substitution Failure Is Not An Error) approach to disable
+   * select functions for attributes, to not support slicing.
+   */
+  template<StorageObjectType U>
+  struct is_dataset
+      : std::integral_constant<bool, (U == StorageObjectType::Dataset)>
+  {
+  };
+
+  /**
+   *  Static assert to enforce the restriction that ReadDataWrapper can only be
+   *  instantiated for OTYPE of StorageObjectType::Dataset and
+   *  StorageObjectType::Attribute but not for other types, e.g., Group or
+   * Undefined.
+   */
+  static_assert(is_allowed_storage_object_type<OTYPE>::value,
+                "StorageObjectType not allowed for ReadDataWrapper");
+
+  // Actual definition of the class
+public:
+  /**
+   * @brief Default constructor.
+   *
+   * @param io The IO object to use for reading
+   * @param dataPath The path to the attribute or dataset to read
+   */
+  ReadDataWrapper(const std::shared_ptr<IO::BaseIO> io,
+                  const std::string& dataPath)
+      : io(io)
+      , dataPath(dataPath)
+  {
+  }
+
+  /**
+   * @brief Function to return the \ref AQNWB::Types::StorageObjectType OTYPE of
+   * the instance
+   */
+  inline StorageObjectType getStorageObjectType() const { return OTYPE; }
+
+  /**
+   * @brief Deleted copy constructor to prevent construction-copying.
+   */
+  ReadDataWrapper(const ReadDataWrapper&) = delete;
```

Comment:
To be honest, I was reading the term "data" as a generic term for anything that represents data. But I can see how a user may misinterpret "data" as equal to "dataset", so probably best to change it. 

How about instead of ``wrapper`` or `data`, we could use the term `DataObject` to refer to datasets/attributes in a more generic way. I.e., rename to `ReadDataObject` and `BaseRecordingDataObject`

If we want to keep the term ``wrapper`` then renaming``ReadDataWrapper` to `ReadWrapper` seems fine.  However, `BaseRecordingWrapper` may be confusing, because it sounds like it is wrapping a whole recording instead of a single dataset that is used for recording. If `BaseRecordingData` is only used for datasets (and not attributes), then I think calling it `BaseRecordingDataset` or `BaseRecordingDatasetWrapper` would work.